### PR TITLE
Parallelize OPTIMIZE FINAL for non-replicated table

### LINKED ISSUE #46770: When I do OPTIMIZE FINAL for non-Replicated table, it should assign merges for all partitions at once.
Currently, it processes them sequentially.

How to reproduce:

Take the `hits` table from ClickBench.
```
CREATE TABLE hits_partitioned AS hits ENGINE = MergeTree PRIMARY KEY (CounterID, EventDate, UserID, EventTime, WatchID) PARTITION BY UserID % 64;
SET max_partitions_per_insert_block = 1000;
INSERT INTO hits_partitioned SELECT * FROM hits;
OPTIMIZE TABLE hits_partitioned FINAL;
```

In the second connection, check the `SELECT * FROM system.merges` query.

### PR DESCRIPTION
**Pull request by aiven-sal**
Close #46770

### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
On  non-replicated tables, OPTIMIZE FINAL schedules merges for all partitions and waits for them to complete

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by robot-ch-test-poll** [comment]
This is an automated comment for commit 831da0b6119cfdeb8812a0e38e2015090fb3f1db with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/49504/831da0b6119cfdeb8812a0e38e2015090fb3f1db/ci_running.html)
The overall status of the commit is 🔴 failure

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🔴 failure</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

File: tests/queries/0_stateless/02293_selected_rows_and_merges.sh [hunk 1]
```diff
@@ -15,7 +15,7 @@ ${CLICKHOUSE_CLIENT} -q "insert into tt select number, 1 from numbers(1e6)"
 
 ${CLICKHOUSE_CLIENT} --optimize_throw_if_noop 1 -q "optimize table tt final" "--query_id=$query_id"
 
-# Here SelectRows and SelectBytes should be zero, MergedRows is 2m and MergedUncompressedBytes is 16m
+# Here for OPTIMIZE FINAL all values are 0, cause merges are executed in background.
```
[the remark this answers is not in the document]
Comment (al13n321) [reply tests/queries/0_stateless/02293_selected_rows_and_merges.sh:18]: The query still waits for the merges to complete, it just doesn't propagate the metrics to query_log. A whole new keyword wouldn't make sense for that.

We really should try to propagate the metrics, but I'm not sure what's a good way to do it.

I guess we should pass the thread group from `CurrentThread::getGroup()` in `merge()` to `MergePlainMergeTreeTask`, then make `transfer_profile_counters_to_initial_query()` add metrics to that thread group's `performance_counters`; either from `task_thread_group->performance_counters` (which I suspect won't work, but worth trying) or from `profile_counters` (like in `write_part_log`). I'm not sure if this should replace the `query_thread_group` or be a different thing.

@CheSema, any pointers?

**Review by al13n321 — CHANGES_REQUESTED** [review]
IIUC, this makes the merge happen in context of a background thread rather than in query context. So, metrics aren't attributed to the query, memory limits are not taken from the query/session/user settings, and log messages won't include the query. Not ideal.

Let's at least propagate the metrics, then it'll be acceptable.

File: tests/queries/0_stateless/02293_selected_rows_and_merges.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (CheSema) [reply tests/queries/0_stateless/02293_selected_rows_and_merges.sh:18]: That changes will break stats on the query.

Stats like this
https://github.com/ClickHouse/ClickHouse/blob/e708f0d9413a71cb788d3d23731fe379f44fd729/tests/queries/0_stateless/02293_selected_rows_and_merges.sh#L16-L20

@al13n321 is right, something has to be done to preserve it.

The most simple way I think is pass `transfer_profile_counters_to_initial_query calback` outside MergePlainMergeTreeTask. It has to add the stats to the query's thread group.

Be aware that this profile_counters does not account all metrics, the comment is actual.
https://github.com/ClickHouse/ClickHouse/blob/e708f0d9413a71cb788d3d23731fe379f44fd729/src/Storages/MergeTree/MergePlainMergeTreeTask.cpp#L34-36

The better approach is to account stats directly to the query's thread group. But I'm not sure that is safe to just reuse its thread group as a thread group for merge. That group consist query context, which has to have limited influence on merging. So for me this is an actual question how to do it elegant.

**Review by CheSema — CHANGES_REQUESTED** [review]
Query stats have to be preserved.

File: tests/queries/0_stateless/02293_selected_rows_and_merges.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (CheSema) [reply tests/queries/0_stateless/02293_selected_rows_and_merges.sh:18]: When `executeHere(task)` called thread group is not shared.

[the remark this answers is not in the document]
Comment (al13n321) [reply tests/queries/0_stateless/02293_selected_rows_and_merges.sh:18]: Ah, so `executeHere()` also creates a new thread group, I see.

**Comment by al13n321** [comment]
Ok, I think this is how exactly to make the profile events propagate correctly:
```
src/Storages/MergeTree/MergePlainMergeTreeTask.h:
@@ -49,6 +49,11 @@ public:
         txn = std::move(txn_);
     }
 
+    void setQueryThreadGroup(ThreadGroupPtr g)
+    {
+        query_thread_group = g;
+    }
+
 private:
     void prepare();
     void finish();
@@ -89,6 +94,7 @@ private:
     MergeTreeTransactionPtr txn;
 
     ProfileEvents::Counters profile_counters;
+    ThreadGroupPtr query_thread_group;
 
     ContextMutablePtr task_context;
```
```
src/Storages/MergeTree/MergePlainMergeTreeTask.cpp:
@@ -108,7 +108,10 @@ void MergePlainMergeTreeTask::prepare()
             std::move(profile_counters_snapshot));
     };
 
-    transfer_profile_counters_to_initial_query = [this, query_thread_group = CurrentThread::getGroup()] ()
+    if (!query_thread_group)
+        query_thread_group = CurrentThread::getGroup();
+
+    transfer_profile_counters_to_initial_query = [this] ()
     {
         if (query_thread_group)
         {
```
Then call `task->setQueryThreadGroup(CurrentThread::getGroup())` in `StorageMergeTree::merge()` in the async case.

Please try it and see if it fixes the two tests.

**Review by al13n321 — APPROVED** [review]

**Comment by al13n321** [comment]
Test failures look related, need to investigate: `test_merge_canceled_by_s3_errors`, `test_memory_limit_success`, `01293_optimize_final_force`, `01825_replacing_vertical_merge`.

**Comment by alexey-milovidov** [comment]
We have had many optimizations for FINAL recently - it will be nice to also have this one finished.

**Comment by alexey-milovidov** [comment]
@aiven-sal Looks abandoned.

**Pull request closed without merging**